tests/test_2025_07_16.py

```diff
@@ -0,0 +1,397 @@
+
+    manager_address = csm.getNodeOperator(node_operator)["managerAddress"]
+
+    remove_keys_tx = csm.removeKeys(node_operator, 0, 1, {"from": manager_address})
+
+    assert "SigningKeyRemoved" in remove_keys_tx.events
+    assert "TotalSigningKeysCountChanged" in remove_keys_tx.events
+    assert "VettedSigningKeysCountChanged" in remove_keys_tx.events
+    assert "DepositableSigningKeysCountChanged" in remove_keys_tx.events
+
+    # Verify charge-related events are NOT emitted when charge == 0
+    assert "KeyRemovalChargeApplied" not in remove_keys_tx.events, "KeyRemovalChargeApplied should not be emitted when charge is 0"
+    assert "BondCharged" not in remove_keys_tx.events, "BondCharged should not be emitted when charge is 0"
+
+
+    """IV. CS Verifier rotation (Vote items #17 - #18)"""
+
+    # Validate CS Verifier rotation
+    assert not csm.hasRole(csm_verifier_role, CS_VERIFIER_ADDRESS_OLD)
+    assert csm.hasRole(csm_verifier_role, CS_VERIFIER_ADDRESS_NEW)
+
+    # =======================================================================
+    # ======================== IPFS & events checks =========================
+    # =======================================================================
+
+    # Uncomment when the IPFS description is defined
+    # metadata = find_metadata_by_vote_id(vote_id)
+    # assert get_lido_vote_cid_from_str(metadata) == IPFS_DESCRIPTION_HASH
+
+    """Validating events"""
+
+    # Validate after voting events count
+    count_vote_items_by_voting_events = count_vote_items_by_events(vote_tx, voting)
+    assert count_vote_items_by_voting_events == EXPECTED_VOTE_EVENTS_COUNT
+    # There's 6 vote items not going through DG + ProposalSubmitted event
+    assert len(voting_events) == EXPECTED_VOTE_EVENTS_COUNT
+
+    # Validate after DG proposal execution events count
+    count_vote_items_by_agent_events = count_vote_items_by_events(dg_tx, agent)
+    assert count_vote_items_by_agent_events == EXPECTED_DG_EVENTS_COUNT
+    # 12 Vote items are going through DG
+    assert len(dg_events) == EXPECTED_DG_EVENTS_COUNT
+
+    # Validate total events count
+    assert count_vote_items_by_voting_events + count_vote_items_by_agent_events == EXPECTED_TOTAL_EVENTS_COUNT, "Incorrect voting items count"
+
+    # Validate PML, ATC, RCC ET Factories removal events
+    validate_evmscript_factory_removed_event(voting_events[0], PML_STABLECOINS_FACTORY)
+    validate_evmscript_factory_removed_event(voting_events[1], PML_STETH_FACTORY)
+    validate_evmscript_factory_removed_event(voting_events[2], ATC_STABLECOINS_FACTORY)
+    validate_evmscript_factory_removed_event(voting_events[3], ATC_STETH_FACTORY)
+    validate_evmscript_factory_removed_event(voting_events[4], RCC_STABLECOINS_FACTORY)
+    validate_evmscript_factory_removed_event(voting_events[5], RCC_STETH_FACTORY)
+
+    # Validate oracle rotation events
+    validate_hash_consensus_member_removed(
```

Comment:
Please add emitted_by check (search for it in other event checkers) for all event checkers here